### PR TITLE
fix(push): added functions for notification

### PR DESCRIPTION
**Pull request by vershwal**
In reference to https://github.com/mozilla/fxa-auth-server/issues/1978

### CODE REVIEW

File: lib/routes/account.js [hunk 1]
```diff
@@ -974,6 +974,31 @@ module.exports = (log, db, mailer, Password, config, customs, checkPassword, pus
...
+            })
+
+            return push.notifyVerifyLogin(accountRecord.uid, devices)
+          }
+        }
+
+        function sendVerifyLogin() {
+          return db.devices(accountRecord.uid).then(
+            function (devices) {
+              //filter the devices variable to devices which can receive a push notification
```
Comment (vershwal) [inline lib/routes/account.js]: @seanmonstar Need to filter this to devices that can receive a push notification (i.e. Firefox 56+ for now). Any pointers on how to do that?

File: lib/push.js [hunk 2]
```diff
@@ -264,6 +275,25 @@ module.exports = function (log, db, config) {
...
+     * Notifies connected devices about a signin confirmation request.
+     *
+    */
+
+    notifyVerifyLogin: function notifyVerifyLogin(uid, devices) {
+      var data = Buffer.from(JSON.stringify({
+        command: PUSH_COMMANDS.VERIFY_LOGIN_NOTIFICATION,
+        data: {
+          title: 'Sign-in Request',
+          body: '',
```
Comment (vershwal) [inline lib/push.js]: Need to take this from https://github.com/mozilla/fxa-content-server-l10n

File: lib/routes/account.js (hunk 1 above)
Comment (seanmonstar) [inline lib/routes/account.js]: The `db.devices` promise will return an array of devices, here's what a device object looks like: https://github.com/mozilla/fxa-auth-server/blob/master/lib/db.js#L451-L466

As for the best way to filter the devices, perhaps others (@philbooth) might have better ideas, but I'd consider checking `device.type` for `"desktop"`, and `device.uaBrowserVersion`... though the version is probably a string, like `"54.0.1"`, so it'd need to be able to compare major versions...

File: lib/push.js (hunk 2 above)
Comment (seanmonstar) [inline lib/push.js]: Right, so we'd want to make use of `gettext` to be able to provide translated notification text. This is the text shown in the notification popup itself. We could either add in `gettext` and the translator directly to this file, or a more hefty change could be to add a `push` file in `senders`... I'm not sure yet which would be the better direction..

**Comment by vershwal** [comment]
Not able to test this locally. For some reason https://github.com/mozilla/fxa-auth-server/pull/2068/files#diff-ab87c506e682e952973178a41792df12R990 always returns an empty array, hence always sends an email confirmation.

I tried signing in from different browser sessions and also used the url http://127.0.0.1:3030/signin?service=sync . 

Any reason why this might be happening?

File: lib/push.js [hunk 3]
```diff
@@ -264,6 +275,25 @@ module.exports = function (log, db, config) {
...
+          url: ''
```
Comment (vershwal) [inline lib/push.js]: Would we be generating a new link over here?

**Comment by philbooth** [comment]
> Not able to test this locally. For some reason https://github.com/mozilla/fxa-auth-server/pull/2068/files#diff-ab87c506e682e952973178a41792df12R990 always returns an empty array, hence always sends an email confirmation.

Hey @vershwal, signing in to a vanilla web session won't work for this because device registration is triggered by browser code. So you need to sign in to Sync using a copy of Firefox that is configured to point at your local dev environment. You can either do that manually in `about:config` or you can create a `user.js` file that looks something like this:

https://pastebin.mozilla.org/9030353

(the preferences are the same either way, I think you'll need all the `.uri` ones, the `.allowHttp` one and the `.urlWhitelist` one)

**Comment by vershwal** [comment]
Tried changing directly in about:config. Looks like I messed things up.
Could you elaborate a little on the user.js method?

**Comment by shane-tomlinson** [comment]
> Could you elaborate a little on the user.js method?

You can also use the [fxa-dev-launcher](https://github.com/vladikoff/fxa-dev-launcher).  Clone the repo, go to the fxa-dev-launcher directory and type:

```bash
FXA_ENV=local npm start
```

**Comment by philbooth** [comment]
> Tried changing directly in about:config. Looks like I messed things up.

Sorry about that! My instructions weren't great.

> Could you elaborate a little on the user.js method?

Yep, so you just want to create a file called `user.js` and put it in your Firefox profile directory. The path for that directory is OS-dependent, so depends what machine you're testing stuff on (more on that [here](http://kb.mozillazine.org/Profile_folder_-_Firefox). Paste the file contents from pastebin, re-open the browser and that should see you right.

Oh, wait...

> You can also use the [fxa-dev-launcher](https://github.com/vladikoff/fxa-dev-launcher).  Clone the repo, go to the fxa-dev-launcher directory and type:
>
> ```bash
> FXA_ENV=local npm start
> ```

...or you can just do what @shane-tomlinson said. He's worth paying attention to!

**Comment by vershwal** [comment]
Thanks for pointing me to fxa-dev-launcher @shane-tomlinson and @philbooth. It solves my problem perfectly.

File: lib/routes/account.js [hunk 4]
```diff
@@ -974,6 +976,33 @@ module.exports = (log, db, mailer, Password, config, customs, checkPassword, pus
...
+          }
+        }
+
+        function sendVerifyLogin() {
+          return db.devices(accountRecord.uid).then(
+            function (devices) {
+              devices = devices.filter(function(device) {
+                return (device.type === 'desktop' && parseFloat(device.uaBrowserVersion) >= 56)
+              })
+              if (devices.length === 0 && !(features.isPushLoginVerifyEnabled(accountRecord.uid, accountRecord.email))) {
```
Comment (seanmonstar) [inline lib/routes/account.js]: We should probably move the feature gate check to the top of this function, that way, if it isn't enabled, we can skip the database read. So, something like:

```js
if (features.isPushLoginVerifyEnabled(accountRecord.uid, accountRecord.email)) {
  return db.devices().then( ... )
} else {
  return sendVerifyLoginEmail()
}
```

File: lib/routes/account.js [hunk 5]
```diff
@@ -974,6 +976,33 @@ module.exports = (log, db, mailer, Password, config, customs, checkPassword, pus
...
+              tokenVerificationId: tokenVerificationId
+            })
+
+            return push.notifyVerifyLogin(accountRecord.uid, devices)
+          }
+        }
+
+        function sendVerifyLogin() {
+          return db.devices(accountRecord.uid).then(
+            function (devices) {
```
Comment (seanmonstar) [inline lib/routes/account.js]: Prefer to use arrow functions as closures. So, `db.devices(accountRecord.uid).then(devices => { ... })`.

File: lib/features.js [hunk 6]
```diff
@@ -27,6 +28,18 @@ module.exports = config => {
...
 
     /**
+     * Return whether or not push notification for sign in confirmation is
+     * enabled for a given user, based on their uid and email address.
+     *
+     * @returns {boolean}
+     */
+     isPushLoginVerifyEnabled(uid, email) {
+      return pushNotificationForSignIn && pushNotificationForSignIn.enabled &&
+        isSampledUser(pushNotificationForSignIn.sampleRate, uid, 'pushNotificationForSignIn' ||
```
Comment (seanmonstar) [inline lib/features.js]: I think this has a parenthesis in the wrong place, should be `isSampledUser(pushNotificationForSignIn.sampleRate, uid, 'pushNotificationForSignIn')`

**Pull request closed without merging**